pkg/code-generator/docgen/cross_project.go

```diff
@@ -0,0 +1,116 @@
+package docgen
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	"github.com/solo-io/solo-kit/pkg/code-generator/parser"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/solo-io/solo-kit/pkg/code-generator/docgen/datafile"
+
+	"github.com/ilackarms/protokit"
+	code_generator "github.com/solo-io/solo-kit/pkg/code-generator"
+	"github.com/solo-io/solo-kit/pkg/code-generator/docgen/options"
+	"github.com/solo-io/solo-kit/pkg/code-generator/model"
+)
+
+func WriteCrossProjectDocs(projectConfigs []*model.ProjectConfig, genDocs *options.DocsOptions, absoluteRoot string, protoDescriptors []*descriptor.FileDescriptorProto) error {
+	if genDocs == nil {
+		return nil
+	}
+	switch genDocs.Output {
+	case options.Hugo:
+		return WriteCrossProjectDocsHugo(projectConfigs, genDocs, absoluteRoot, protoDescriptors)
+	default:
+		return nil
+	}
+}
+
+// Hugo docs require a datafile that is used by shortcodes
+func WriteCrossProjectDocsHugo(projectConfigs []*model.ProjectConfig, genDocs *options.DocsOptions, absoluteRoot string, protoDescriptors []*descriptor.FileDescriptorProto) error {
+	// only write the data file if a data dir is specified
+	hugoOptions := genDocs.HugoOptions
+	if hugoOptions == nil {
+		return nil
+	}
+	if hugoOptions.DataDir == "" {
+		return nil
+	}
+	hugoPbData := datafile.HugoProtobufData{
+		Apis: make(map[string]datafile.ApiSummary),
+	}
+
+	for _, projectConfig := range projectConfigs {
+		project, err := parser.ProcessDescriptors(projectConfig, projectConfigs, protoDescriptors)
+		if err != nil {
+			return err
+		}
+		// Collect names of files that contain resources for which doc gen has to be skipped
+		skipMap := make(map[string]bool)
+		for _, res := range project.Resources {
+			if res.SkipDocsGen {
+				skipMap[res.Filename] = true
+			}
+		}
+		protoFiles := protokit.ParseCodeGenRequest(project.Request)
```

Comment:
Similar in spirit to the above nit: in `parser.parser.go:37` we already call this function with the same argument (the `req` that is used there is then set as the `Project.Request` that you use here). If we update `model.Project` to have a field like this

```go
Descriptors []*FileDescriptor
```
we could reuse the result of the previous function call and this line would become:

```go
protoFiles := project.Descriptors
```